### ISSUE TITLE
Recurring option and mass add? 

### ISSUE THREAD

**Issue by Ketrel** [brarcher/budget-watch#99]
I was thinking along the lines of

1. Checkbox on a transaction "recurring".  
2. Menu item on transaction screen "Add recurring"  
3. When clicked, pops up a date range selector, and then for any items with "recurring" checkbox selected in the range, duplicates them but with the month incremented?

**Comment by brarcher** [comment]
Thanks for the suggestion.

I've seen a recurring feature in another budgets app I used in the past. The creating of entries on a recurring basis of that app seemed odd to me, so I did not investigate adding the feature to Budget Watch. I'm not opposed to having such a feature, but I may ask for assistance from someone taking it on who is interested and has time.

In the meantime, the following is my understanding of your use case and how it might work under the covers. Let me know if I am missing anything or if you disagree.

1) Definitions
- A _transaction_ is an entry which defines a debit or credit. It has a name, value, budget, and date. When a transaction is added, its value will be calculated into its related budget.
- A _recurring transaction_ is an entry which defines a future debit or credit. It has a name, value, budget, recurring schedule, and time window over which it is valid. Its value is not calculated into its related budget. A recurring transaction results in the creation of transactions in the future, which are calculated into its related budget.
2) Adding a recurring transaction
- When adding a transaction there would be an option for it to be recurring. If checked, an option to see how often to add the transaction would appear. One can have the transaction reoccur every X days, or a specific day of the month. Further, an option date selection will be visible, to show in what time window the transaction can reoccur. If the transaction is added with the recurring option selected, the added entry is a recurring transaction.
- Once added, if the recurring transaction is valid to reoccur at that moment, a new transaction will be created immediately.
2) Recurring
- When the app is loaded*, a check for all recurring transactions is performed. If a recurring transaction is due, a new transaction is added for each recurring due.
* (The definition of 'loaded' is ambiguous. The app could be active for an indefinite amount of time, in which case the checking will not be triggered. Determining what triggers the recurring check will need to be investigated).
3) Modifying/Deleting a recurring transaction
- If an existing recurring transaction is modified all previously generated transactions do not change. However, future generated transactions will reflect the updates.
- If an existing recurring transaction is deleted all previously generated transactions remain. No future transaction will be added due to the removed recurring transaction.

**Comment by Ketrel** [comment]
I had originally typed that from a phone, so I didn't get as in depth as I should have.
I was thinking something more manual than that.

1. When adding/editing a transaction, it would have a checkbox for 'recurring'.
2. If checked, it would have something like a dropdown with options like 'Weekly, Bi-Weekly, Monthly, Quarterly, Semi-Annually, Annually"
3. A menu option to 'Add Recurring Transactions (For Period)'

When selected, that menu option would 
1. Open a calendar selector to choose a date (by default would show the current date range that the budget is set to).
2. Look back at transactions over each of the (pre-defined) ranges
3. Add a duplicate transaction with everything identical except the date is updated (if not already existing).

So for example, if I have a transaction on March 1st called "Dog Food Order", with recurring checked, and 'Monthly', and I choose April 1 - April 30 for my range when I choose to add recurring, it would

1. Look at the annually within April 1 - 30 from a year ago for anything set to recur annually and duplicate it
2. Look at October 1 - 30  for anything set to recur semi-annually and duplicate it
3. Look at January 1 - 30 for anything set to recur quarterly and duplicate it 
4. Look at March 1 - 30 for anything set to recur monthly and duplicate it  
  (Here is where it would create a duplicate "Dog Food Order" transaction on April 1, set to recur monthly. 
5. etc for weekly and bi-weekly

So, I wasn't thinking an automated add, but just the ability to basically script duplicating the transactions.

**Comment by brarcher** [comment]
So, if I understand correctly, the feature would result in several transactions being added immediately for different dates, rather than automatically adding transactions in the future. I did not consider something like that. Seems interesting, and less complicated than my idea. (:

**Comment by Ketrel** [comment]
Exactly.  It's basically a glorified copy/paste operation except it automates which ones to copy and what date to paste them to based on those options.  

Plus it would be easy to end a recurring transaction by editing the last instance of it and unchecking 'recurring' since when you run it the next month, it wouldn't pick up that one as recurring and therefore wouldn't copy it.

**Comment by MarimbaMaurice** [comment]
I don't really get the script duplicating transaction.
What's the advantage of this over the automatic add?

**Comment by Ketrel** [comment]
Nothing has to run in the background or be scheduled, plus it's an option, not automatically added.
